sycl/include/CL/sycl/ONEAPI/reduction.hpp

```diff
@@ -1004,6 +1056,461 @@ reduAuxCGFunc(handler &CGH, size_t NWorkItems, size_t MaxWGSize,
+      if (LID < CurStep) {
+        // LocalReds[LID] = BOp(LocalReds[LID], LocalReds[LID + CurStep]);
+        reduceReduLocalAccs(LID, LID + CurStep, LocalAccsTuple, BOPsTuple,
+                            ReduIndices);
+      } else if (!UniformPow2WG && LID == CurStep && (PrevStep & 0x1)) {
+        // LocalReds[WGSize] = BOp(LocalReds[WGSize], LocalReds[PrevStep - 1]);
+        reduceReduLocalAccs(WGSize, PrevStep - 1, LocalAccsTuple, BOPsTuple,
+                            ReduIndices);
+      }
+      NDIt.barrier();
+      PrevStep = CurStep;
+    }
+
+    // Compute the partial sum/reduction for the work-group.
+    if (LID == 0) {
+      size_t GrID = NDIt.get_group_linear_id();
+      // If there is only one work-group, then the original accessors need to be
+      // updated, i.e. after the work in each work-group is done, the work-group
+      // result is added to the original value of the read-write accessors or
+      // USM memory.
+      std::conditional_t<IsOneWG, IsRWReductionPredicate,
+                         EmptyReductionPredicate>
+          Predicate;
+      auto RWReduIndices =
+          filterSequence<Reductions...>(Predicate, ReduIndices);
+      writeReduSumsToOutAccs<UniformPow2WG>(
+          GrID, WGSize, (std::tuple<Reductions...> *)nullptr, OutAccsTuple,
+          LocalAccsTuple, BOPsTuple, ReduIndices, RWReduIndices);
+    }
+  });
+}
+
+template <typename KernelName, typename KernelType, int Dims,
+          typename... Reductions, size_t... Is>
+void reduCGFunc(handler &CGH, KernelType KernelFunc,
+                const nd_range<Dims> &Range,
+                std::tuple<Reductions...> &ReduTuple,
+                std::index_sequence<Is...> ReduIndices) {
+  size_t NWorkItems = Range.get_global_range().size();
+  size_t WGSize = Range.get_local_range().size();
+  size_t NWorkGroups = Range.get_group_range().size();
+
+  bool Pow2WG = (WGSize & (WGSize - 1)) == 0;
+  bool HasUniformWG = Pow2WG && (NWorkGroups * WGSize == NWorkItems);
+  if (NWorkGroups == 1) {
+    if (HasUniformWG)
+      reduCGFuncImpl<KernelName, true, true>(CGH, KernelFunc, Range, ReduTuple,
+                                             ReduIndices);
+    else
+      reduCGFuncImpl<KernelName, false, true>(CGH, KernelFunc, Range, ReduTuple,
+                                              ReduIndices);
+  } else {
+    if (HasUniformWG)
+      reduCGFuncImpl<KernelName, true, false>(CGH, KernelFunc, Range, ReduTuple,
+                                              ReduIndices);
+    else
+      reduCGFuncImpl<KernelName, false, false>(CGH, KernelFunc, Range,
+                                               ReduTuple, ReduIndices);
+  }
+}
+
+// The list of reductions may be empty; for such cases there is nothing to do.
+// This function is intentionally made template to eliminate the need in holding
+// it in sycl library, what would be less efficient and also would create the
```

Comment:
```suggestion
// it in SYCL library, what would be less efficient and also would create the
```

Reply:
Fixed in f0d0716cc063bfcdf3ac9a90fd052ce539699a9c